winter_messaging_transactional/tests/tests/test_consumer.py

```diff
@@ -0,0 +1,249 @@
+def test_consume_with_error(event_consumer, rabbit_url, event_processor, event_publisher, consumer_dao, session, can_be_handled_on_retry):
+    event_id = 1
+    payload = 'consume_with_error'
+    event = RetryableEvent(id=event_id, payload=payload, can_be_handled_on_retry=can_be_handled_on_retry)
+
+    # Act
+    with event_consumer('consumer_with_error'):
+        event_publisher.emit(event)
+        session.commit()
+        sleep(10)  # By default, there are 3 attempts to handle the event
+
+    # Assert
+    inbox_messages = read_all_inbox_messages(session)
+    assert len(inbox_messages) == 1
+
+    event_message = inbox_messages[0]
+    assert event_message['consumer_id'] == 'consumer_with_error'
+    assert event_message['name'] == 'RetryableEvent'
+    assert event_message['counter'] == (2 if can_be_handled_on_retry else 3)
+    assert event_message['received_at']
+    if can_be_handled_on_retry:
+        assert event_message['processed_at']
+    else:
+        assert event_message['processed_at'] is None
+
+    consumed_event_1 = consumer_dao.find_by_id(event_id)
+    consumed_event_2 = consumer_dao.find_by_id(event_id + 1)
+
+    connection = create_rabbitmq_connection(rabbit_url)
+    channel = connection.channel()
+    ack, properties, message_payload = channel.basic_get(queue='winter.dead_letter_queue', auto_ack=True)
+
+    if can_be_handled_on_retry:
+        assert ack is None
+        assert consumed_event_1
+        assert consumed_event_2
+        assert consumed_event_1['payload'] == payload
+        assert consumed_event_2['payload'] == payload
+    else:
+        assert consumed_event_1 is None
+        assert consumed_event_2 is None
+
+        assert ack
+        assert ack.exchange == 'winter.dead_letter_exchange'
+        assert ack.routing_key == 'sample-topic.RetryableEvent'
+        assert properties.type == 'RetryableEvent'
+        assert message_payload == b'{"id": 1, "payload": "consume_with_error", "can_be_handled_on_retry": false}'
+
+
+def test_deduplication_on_consumer_side(event_consumer, event_processor, injector, consumer_dao, session, rabbitmq_client):
+    event_id = 1
+    consumber_id = 'consumer_correct'
+    topic = 'sample-topic'
+    message_id = uuid4()
+
+    inbox_message_dao = injector.get(InboxMessageDAO)
+    inbox_message = InboxMessage(
+        id=message_id,
+        consumer_id=consumber_id,
+        name='SampleEvent'
+    )
+    inbox_message_dao.upsert(inbox_message)
+    inbox_message_dao.mark_as_processed(message_id, consumber_id)
+    session.commit()
+
+    duplicated_message = OutboxMessage(
+        message_id=message_id,
+        topic=topic,
+        type='SampleEvent',
+        body=json.dumps(dict(payload='payload'), ensure_ascii=False, cls=JSONEncoder),
+    )
+
+    # Act
+    with event_consumer(consumber_id):
+        rabbitmq_client.publish(duplicated_message, get_exchange_name(topic))
+        sleep(5)
```

Comment:
I suggest to replace sleeps with polling in a loop, benefits
- quicker tests (5 seconds is definitely too much to process a single message)

Reply:
Can we use wait_for_result here?